hf_olmo/convert_olmo_to_hf.py

```diff
@@ -60,46 +101,154 @@ def write_tokenizer(checkpoint_dir: str):
+    tokenizer_name_or_path = str(tokenizer_name_or_path or conf["tokenizer"]["identifier"])  # pyright: ignore
 
-    convert_checkpoint(local_model_path)
-    return local_model_path
+    try:
+        Tokenizer.from_pretrained(tokenizer_name_or_path)
+    except Exception as e:
+        logger.error(f"Error loading tokenizer: {e}")
+        raise e
 
+    conf["tokenizer"]["identifier"] = tokenizer_name_or_path  # pyright: ignore
+
+    if tokenizer_name_or_path == "allenai/gpt-neox-olmo-dolma-v1_5":
+        conf["model"]["eos_token_id"] = 50279  # pyright: ignore
 
-def fix_bad_tokenizer(checkpoint_dir: str):
-    path = os.path.join(checkpoint_dir, "config.yaml")
-    conf = om.load(path)
-    conf["tokenizer"]["identifier"] = "allenai/gpt-neox-olmo-dolma-v1_5"
-    conf["model"]["eos_token_id"] = 50279
     om.save(conf, path)
 
 
+def download_s3_directory(bucket_name: str, prefix: str, local_dir: str, ignore: str | None = None):
+    # Create S3 client
+    s3_client = boto3.client("s3")
+
+    re_ignore = re.compile(ignore) if ignore else None
+
+    # List objects within the given prefix
+    paginator = s3_client.get_paginator("list_objects_v2")
+    pages = paginator.paginate(Bucket=bucket_name, Prefix=prefix)
+
+    # Create a list to hold all the files to download
+    files_to_download = []
+    for page in pages:
+        for obj in page.get("Contents", []):
+            if re_ignore and re_ignore.search(obj["Key"]):
+                continue
+            files_to_download.append(obj["Key"])
+
+    # Initialize the progress bar
+    with tqdm(total=len(files_to_download), desc="Downloading files") as pbar:
+        for s3_key in files_to_download:
+            # Construct the full local path
+            local_file_path = os.path.join(local_dir, os.path.relpath(s3_key, prefix))
+            local_file_dir = os.path.dirname(local_file_path)
+
+            # Ensure local directory exists
+            if not os.path.exists(local_file_dir):
+                os.makedirs(local_file_dir)
+
+            # Download the file
+            s3_client.download_file(bucket_name, s3_key, local_file_path)
+
+            # Update the progress bar
+            pbar.update(1)
+
+
+@contextmanager
+def make_local_checkpoint(checkpoint_dir: str) -> Generator[str, None, None]:
+    parsed_dir = urlparse(checkpoint_dir)
+
+    assert parsed_dir.scheme in ["s3", ""], "Only s3 and local paths are supported."
+
+    if os.path.exists(checkpoint_dir):
+        yield checkpoint_dir
+        return
+
+    temp_dir = os.path.join(tempfile.gettempdir(), md5(checkpoint_dir.encode()).hexdigest())
+    if os.path.exists(temp_dir):
+        yield temp_dir
+        return
+    try:
+        os.makedirs(temp_dir, exist_ok=True)
+        print(f"Downloading checkpoint to {temp_dir}...")
+        download_s3_directory(
+            bucket_name=parsed_dir.netloc,
+            prefix=parsed_dir.path[1:],
```

Comment:
Why do you need to remove the first letter? If it's to get rid of `/`, then `lstrip("/")` might indicate your intent more clearly

Reply:
chatgpt code 🤪